Optimize and test MemoizedSeriesIterator

leizor opened the pull request:
<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

https://github.com/prometheus/prometheus/issues/11274

This PR makes a couple of changes to `MemorizedSeriesIterator`:
- Remove support for `AtHistogram()`; this iterator is only used in the PromQL engine which only works with float histograms.
- Automatically convert integer histograms to float histograms when calling `AtFloatHistogram()` (I'm not sure about this change; it looks like the iterator can point to an integer histogram but we want an auto-converted float histogram? Or is it that the iterator can never point to an integer histogram? I'm guessing the former, but let me know if I'm wrong!)

This PR also adds tests testing the iterator's support native histograms.

leizor commented:
I also noticed slightly different behavior between `Seek` and `Next` for this iterator.

Given a series of samples: 

```
{ts: 1, v: 1}
{ts: 2, v: 2} <- iterator starts here 
{ts: 10, v: 10}
{ts: 11, v: 11}
```

If you call `Next` and advance the iterator to the next sample, `PeekPrev` will return `{ts: 2, v: 2}` and `ok=true` whereas if you call `Seek(10)` `PeekPrev` will return `ok=false`.  Is `Next` supposed to invalidate the previous sample when we jump a timestamp > `b.delta` like `Seek` does?

fpetkovski commented on storage/memoized_iterator.go:
```diff
@@ -108,15 +103,14 @@ func (b *MemoizedSeriesIterator) Next() chunkenc.ValueType {
 		b.prevValue = 0
-		b.prevTime, b.prevHistogram = b.it.AtHistogram()
-		_, b.prevFloatHistogram = b.it.AtFloatHistogram()
+		ts, h := b.it.AtHistogram()
```
If we are removing support for `AtHistogram`, can we call `b.it.AtFloatHistogram` directly to avoid allocating data twice? We will never need the original histogram and we can decode to a float histogram directly.

leizor replied:
I wasn't sure if we could depend on the underlying iterator being able to perform the float histogram conversion directly but looking closer that does seem to be the case--I'll make the change!

leizor replied:
Made the change as well as adding a comment clarifying the assumption.

fpetkovski commented on storage/memoized_iterator.go:
```diff
@@ -132,13 +126,17 @@ func (b *MemoizedSeriesIterator) At() (int64, float64) {
+// AtHistogram is not supported by this iterator.
 func (b *MemoizedSeriesIterator) AtHistogram() (int64, *histogram.Histogram) {
-	return b.it.AtHistogram()
+	panic("AtHistogram is not supported by this iterator.")
```
If `MemoizedSeriesIterator` is used only in PromQL, maybe we could introduce a dedicated interface which does not expose `AtHistogram` at all.

It also feels strange that `Next` can return `ValueTypeHistogram` but calling `AtHistogram` would panic.

leizor replied:
Hm, that's a good point--I'll look into it

leizor commented:
After looking a bit more closely at this, I've made this PR a bit more opinionated. It looks like this iterator is only used [here](https://github.com/prometheus/prometheus/blob/0d049feac760444cbbafadbe8794843a6f47dba4/promql/engine.go#L1836-L1869) and it's not using it as the `chunkenc.Iterator` interface but rather a pointer to the struct.

I'm proposing we make `MemoizedSeriesIterator` no longer implement `chunkenc.Iterator`:
- I've made renamed `Next` to `next`; this gets around that difference in invalidating previous samples behavior I pointed out earlier by preventing external use.
- I removed `AtHistogram` and made calls to `Seek` and `next` only return `ValNone`, `ValFloat`, and `ValFloatHistogram`.
- I removed `AtT`; it's not used.

I don't know if this is ok to do as this makes it a breaking change...anyways, let me know!

beorn7 commented on storage/memoized_iterator.go:
```diff
@@ -132,21 +128,15 @@ func (b *MemoizedSeriesIterator) At() (int64, float64) {
 func (b *MemoizedSeriesIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
+	if b.valueType == chunkenc.ValHistogram {
+		ts, h := b.it.AtHistogram()
+		return ts, h.ToFloat()
```
The contract is indeed that `AtFloatHistogram` works if `chunkenc.ValHistogram` is returned, see https://github.com/prometheus/prometheus/blob/0ab95536115adfe50af249d36d73674be694ca3f/tsdb/chunkenc/chunk.go#L130

beorn7 replied:
In different news, `b.valueType` will never be `chunkenc.ValHistogram`, will it?

I think we can really just forward to `b.it.AtFloatHistogram` without further ado.

leizor replied:
Good point - fixed!

beorn7 commented on storage/memoized_iterator.go:
```diff
@@ -20,7 +20,9 @@ import (
-// MemoizedSeriesIterator wraps an iterator with a buffer to look back the previous element.
+// MemoizedSeriesIterator wraps an iterator with a buffer to look back the previous element. This iterator regards
+// integer histograms as float histograms. Calls to Seek() will never return chunkenc.Histogram and it expects
+// that the chunkenc.Iterator provided to it can perform the conversion directly with a call to AtFloatHistogram().
```
Which is the contract anyway, see https://github.com/prometheus/prometheus/blob/0ab95536115adfe50af249d36d73674be694ca3f/tsdb/chunkenc/chunk.go#L130

beorn7 replied:
Maybe note here that `MemoizedSeriesIterator` deliberately does not implement `chunkenc.Iterator`.

beorn7 commented on storage/memoized_iterator.go:
```diff
@@ -85,13 +82,15 @@ func (b *MemoizedSeriesIterator) Seek(t int64) chunkenc.ValueType {
 		b.valueType = b.it.Seek(t0)
 		if b.valueType == chunkenc.ValNone {
 			return chunkenc.ValNone
+		} else if b.valueType == chunkenc.ValHistogram {
+			b.valueType = chunkenc.ValFloatHistogram
 		}
```
We generally (and as of recently ;) prefer `switch` instead of `else if` constructs:

```suggestion
		switch b.ValueType {
			case chunkenc.ValNone:
				return chunkenc.ValNone
			case chunkenc.ValHistogram:
				b.valueType = chunkenc.ValFloatHistogram
		}
```

beorn7 commented on storage/memoized_iterator.go:
```diff
@@ -101,29 +100,26 @@ func (b *MemoizedSeriesIterator) Seek(t int64) chunkenc.ValueType {
 }
 
 // Next advances the iterator to the next element.
```
Perhaps add a note why `next` is un-exported and what the difference in behavior is.

beorn7 replied:
With your note in, I now get second thoughts about the unexporting. I think `next` behaves fine and isn't violating any contract of the original `Next`. The specifics about the previous samples are really specific to the `MemoizedSeriesIterator`. Plus, we aren't even implementing a normal `chunkenc.Iterator` anyway (because other methods got removed). I now think we can just leave `Next` as Next`, maybe documenting that the `PeekBack` thing only works with `Next` and not with `Seek`. @codesome WDYT?

leizor replied:
I agree that it didn't violate the contract of the original `next`. I was thinking the reason to not export it is because it might be surprising that it doesn't work correctly with `PeekBack` but just having that comment could also work.

beorn7 replied:
Isn't it actually the other way around? _Only_ after a call of `Next` (or `next` ;) `PeekBack` will work correctly. After `Seek`, `PeekBack` was never supposed to work.

And that behavior didn't change with this PR, did it?

I would really make `next` `Next` again, and if at all, add to the doc comment of `Seek` that, after it, `PeekBack` will return false.

beorn7 replied:
Yeah, PeekPrev is only ever used after Seek. Since Seek uses Next internally, the current behavior might be needed.

This iterator has a weird semantics… I think indeed we better don't touch anything here. A more comprehensive explanation might be good at some time in the future so that not everyone has to wrap their head around this semantics again.

beorn7 commented on storage/memoized_iterator_test.go:
```diff
@@ -16,25 +16,36 @@ package storage
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/tsdbutil"
+
+	"github.com/stretchr/testify/require"
```
We usually do import packages outside of github.com/prometheus/prometheus in a block _before_ the block with imports from github.com/prometheus/prometheus (as it was before).

beorn7 reviewed (changes requested):
Cool, looks good, and almost there.

Most comments just style and doc comment nits. The only substantial change is to simplify the `AtFloatHistogram` implementation.

beorn7 commented:
On the code level, this PR is OK once the open comments are addressed, but I would like to get a high-level confirmation from @codesome that it is OK to let `MemoizedSeriesIterator` _not_ implement `chunkenc.Iterator`.

beorn7 reviewed (approved):
Thank you very much. I'll merge now, but @codesome if you could do the high-level check eventually (is it OK to not implement the iterator interface), that would be cool.
